### ISSUE TITLE
Not getting comics after #1662

### ISSUE THREAD

**Issue by justinjustin7** [mamaral/xkcd-Open-Source#39]
Title says it all. Haven't been getting updates for about a week now.

**Comment by mamaral** [comment]
@fergusean

**Comment by fergusean** [comment]
will check tomorrow. there must be another breaking comic.

**Comment by Vintodrimmer** [comment]
Same problem.

**Comment by The-Penultimate-Defenestrator** [comment]
I have this too. The problem was the Garden comic, I think. It was interactive, and the app broke after this. It was released April 4, I believe.

**Comment by mamaral** [comment]
Our temporarily solution is to manually insert an image into the db for breaking comics. I propose one of two solutions:

1) Create an image that we insert any time this issue occurs, that basically tells users to go to the web to check out the interactive comic.

2) Add some metadata to comics that would allow us to open a web view to the xkcd website at that comic.

**Comment by mamaral** [comment]
I'm not entirely sure if any of those comics would work in a web view on an iOS device, I'd have to try it in Safari. I'm at work now so won't be looking into this until tonight at the earliest.

**Comment by The-Penultimate-Defenestrator** [comment]
I'll check, I'm on my phone now.

**Comment by The-Penultimate-Defenestrator** [comment]
Garden seems to work fine. ![](http://i.imgur.com/nNvXfMl.jpg)

What were the other ones? I imagine some older ones might use Flash instead of a HTML5  `<canvas>`

**Comment by mamaral** [comment]
We've been keeping track as we find them https://github.com/mamaral/xkcd-Open-Source/issues/2

With that in mind, I vote that if the image is of some other format we return a some indicator that we should show the comic in a web view rather than the native comic view controller, so the clients can handle those a bit more elegantly. We could add the logic to the server to flag future comics and give us back the URL, and manually fix the existing broken comics.

**Comment by NguyenDytrich** [comment]
Most of the interactive comics in #2 have image links despite being HTML-based or canvas based, though. I agree with taking the web view approach, but how will we flag a comic as interactive?

**Comment by The-Penultimate-Defenestrator** [comment]
Note that the interactive comics do still provide a link to a static image,
such as http://imgs.xkcd.com/comics/frequency.png. This could be a simple
solution to handling interactive comics.

I haven't looked at the code yet (it may be above my level) but I will
likely be able to contribute to the server-end of things. However, I don't
know Objective C or Swift, so I can't help with the app side. That's the
only reason I think a solution like replacing with static images would be
ideal.

**Comment by mamaral** [comment]
When generating the comics, if the image doesn't have a recognized extension, .png, jpg, etc., we generate the URL and flag the comic. That would solve the cases where its not even an image, like this one http://xkcd.com/1608/info.0.json. The others that are interactive but include image files, like the Emojic 8 Ball (http://xkcd.com/1525/info.0.json) will likely have to be manually done for now? I can't think of any way the server would know - assuming the data might be the same as a normal comic in some cases.

**Comment by The-Penultimate-Defenestrator** [comment]
I don't think there is a way to tell. There might be certain things you
could do, like checking the page HTML for JavaScript that isn't normally
there. This is liable to break though as he adds new things to the website.

**Comment by mamaral** [comment]
Yeah, IMHO it's not worth doing some external hacking that might not work 100% anyway. I think my solution as proposed above is probably the way to go, for now at least. Handle on the server if the image isn't an image based on the extension, and manually flag anything else we find that is interactive.

**Comment by The-Penultimate-Defenestrator** [comment]
Ok. What are the interactive comics that *don't* provide an image, just so
that I can take a look?

**Comment by mamaral** [comment]
Here's one - http://xkcd.com/1608/info.0.json

**Comment by mamaral** [comment]
@fergusean Can we add an `isInteractive` boolean and a `webURL` string to the comics? That would be enough for clients to handle showing them in a web view, which I could build into the next release.

The server side of things would need to do the following:

1) If the extension is not an image extension, flag the comic as interactive and create the web URL.
2) Migrate all existing comics data to default to false on the interactive flag.
3) Manually flag the known interactive comics as interactive and generate their web URLs.

How's that sound?

**Comment by The-Penultimate-Defenestrator** [comment]
I could try to do that. I might not get to it for a while though.

**Comment by NguyenDytrich** [comment]
We should be able to use `[Comic generateShareUrl]` to generate the url for the comic. I don't know if we would have to migrate all existing comics to a default flag; the metadata is stored as JSON and `self.isInteractive = dictionary[kInteractiveKey] ?: NO;` should work without migrating any data, but I'm not super fluent in Objective-C.

**Comment by mamaral** [comment]
You are correct in that it would work if we didn't have the flags on existing comics, as if a field isn't present in a dictionary the boolValue will default to false, but I would rather have an explicit false and all the data be consistent.

**Comment by The-Penultimate-Defenestrator** [comment]
Looks like it's fixed now. I just got a notification for Monday's comic.

**Comment by mamaral** [comment]
Indeed - looks like everything is patched up for the moment. I'm working on v2.2 that includes handling interactive comics, will need to talk to @fergusean about data model changes from the API.

**Comment by fergusean** [comment]
sorry for the delay. I've been on a work trip, and internet has been spotty. manually injected 1663 with a screenshot from the site into the DB.

@mamaral that sounds good. are you going to insert a standard placeholder for interactive comics?

**Comment by mamaral** [comment]
The hard part is that some of the interactive comics have images, just static ones, and some aren't images at all. It sucks to replace the actual images with placeholders, but it also sucks to see white if there is no image...

Not sure if we want a client-side or server-side solution to the image problem... I might just photoshop up a different placeholder for interactive comics and force the tiles to use that on the comic list, and when clicked we open the web view that will (hopefully) be able to be interacted with.

**Comment by controversial** [comment]
Since this issue is fixed, should it be closed, and we could discuss the displaying interactive comics in another issue?
